jython-3-roadmap.md

```diff
@@ -0,0 +1,293 @@
+
+1. Definition of classes, members and methods using annotations in Java.
+(Something like the Jython 2 exposer but less opaque, documented,
+and simplified using `MethodHandle`.)
+
+### Experiment with Object
+
+Consider the advantages to performance,
+and to the transparency of Java integration,
+of making every `Object` a Python object.
+Explore the idea of "acceptable implementations"
+of common built-in types to allow e.g. `String` to be a `str`.
+Experiment with ``CallSite`` as a consumer of the `MethodHandle`
+already in slots.
+
+Resolve the `PyObject` vs `Object` dilemma.
+
+### Java Integration
+
+Approach to and basic implementation of
+treating Java objects as Python objects,
+having a Python type related to their Java class,
+when they have not been specifically identified (built-ins).
+Performance micro-benchmarks modelling code compiled to Java. 
+
+### Module Import
+
+1. Outline architecture for modules and importers,
+giving special attention to the semantics of Java packages and modules.
+Advances in the module concept in Python should allow us
+to avoid some of the special cases and thrashing around we find in Jython 2.
+
+1. Rudimentary forms of `sys`, `io`.
+Subsequently, objects to be added as needed.
+
+1. Implement import mechanism closely following CPython.
+
+1. Use custom finders (probably) to import objects from Java.
+
+
+### Compiler
+
+1. Further selected `stdlib` modules as necessary in the compiler.
+
+1. AST classes generated from Python ASDL.
+Generated classes are Python objects in an `ast` module.
+(Question: should they be generated in Java? With ANTLR?)
+
+1. Compiler from Python source to AST,
+probably using the PEG parser.
+(If adopting PEG, compile it with CPython and run it with Jython.)
+
+1. Compiler from AST to CPython byte code:
+using the version in Python if possible (compiled with CPython).
+Otherwise, follow CPython implementation in Java.
+(There is no CPython byte code compiler in Jython 2 legacy.)
+
+### Jython Command
+
+1. Jython 3 console command as a Java application
+built over the library.
+
+1. Means to invoke Jython on all major OSes.
+
+### Python `stdlib`
+
+Progressively introduce the `stdlib` and its regression test suite.
+
+The CPython regression test suite is hugely useful
+for driving our conformance and completeness,
+but the test process itself relies on a large proportion
+of the language and `stdlib` working already.
+
+## Implementation Notes
+
+A few principles, some drawn from discussions on `jython-dev` or off-list.
+
+* Adopt or write a higher proportion of modules in Python vs Java,
+  than is the case in Jython 2.
+  (Decent but not high performance interpreter is a pre-requisite.)
+  * Only performance-critical modules are hand-crafted in Java.
+  * Modules that take advantage of Java libraries call them from Python.
+  * Enthusiasm for writing in Java should be directed to re-implementing
+    popular libraries that broaden Jython use (`numpy`, etc.).
```

Comment:
Maybe mention ctypes, cffi and a cython-backend here.

Reply:
I can mention them, but what would we like actually to say?

More broadly, I know calling modules written in  C would be useful, but until ideas of the pure Java core settle, I find it difficult to say what's involved. So not MVP, but interested in your thoughts now and when that settling has happened.

Reply:
Calling C by pure Java code won't be feasible until Project Panama lands. Currently one inherently has to write platform dependent C glue code or use something like JNA or JNR which is just the same with the only difference is that it was written by others. ctypes, cffi and cython are JNA/JNR-like libs for Python. Having Jython-versions would enable tons of Python modules at a time. This is still not sufficient for NumPy and SciPy but is a requirement. All three are used by NumPy and SciPy. So if you want to support scientific code, getting ctypes, cffi and cython working would be the first step.
That said, each of these is a huge project on its own. So it is likely a better approach to support [https://github.com/hpyproject/hpy](HPy) and hope that these libraries will be made HPy compatible as well. I know that cython announced HPy support. But then, supporting HPy is not feasible with pure Java code either.